### DISCUSSION TITLE
Metrics aggregation <MISSING VALUE> and and other questions

### DISCUSSION THREAD

**Discussion by nvasoft** [jertel/elastalert2#785] [Q&A] [answered]
Hello,

I need to process a huge amount of events in large amount of a time (day or week).
I used the Cardinality rule, after it began to give an error:  `ProcessController: WARNING:apscheduler.scheduler:Execution of job "ElastAlerter.handle_rule_execution…" skipped: maximum number of running instances reached (1)` I began to use the Metrics aggregation rule with `metric_agg_type: cardinality.` Metric agg makes different query so it is lighter for elastalert.
I ran into several problems:
1.	fields are not displayed in `alert_text_args: `<missing value>, except those described in query key.
let's say:
```
query_key:
  - source.ip
  - threat.name
```
output:
```
Log.level: <missing value>
threat.name: threatname1
source.ip 1.2.3.4
```
2.	How to increase the calculation window (day or week)? By default this is buffer_time in `config.yaml`.
If I separately use `buffer_time: days: 1` in the rule itself, will it affect other rules?
3.	And I need to count the number of events for the aggregation schedule, how many were flow_id for a given flow in query_key. I used in the Cardinality rule `len(self.rule["type"].cardinality_cache[match[...]])`.
What should I use now to count?

**Comment by jertel** [comment]
> 1. fields are not displayed in alert_text_args: , except those described in query key.

Correct. Elasticsearch only returns counts with aggregations, not the entire document. That wouldn't make much sense since it wouldn't know which document to return, as the aggregation was across many documents.

> 2. If I separately use buffer_time: days: 1 in the rule itself, will it affect other rules?

No. Only configurations specified in the global elastalert config YAML will apply to all rules. Configurations placed inside of a specific rule YAML will not leak into other rules.

> 3. What should I use now to count?

You are attempting something unsupported with ElastAlert 2. I tried to give guidance in your previous thread but I'm unable to devote more time to this. You may need to find a software developer to assist you.

**Reply by nvasoft** [reply]
@jertel thank you very much for your support of this project and the help you provide.
For calculation I used `timeframe: days: 7` in cardinality rule.. So here I need to use `buffer_time: days: 7`, inside the rule YAML?
The rest I'll try to figure out myself. If I find an answer, I'll post in the thread. Perhaps the value of self.metric_key will help count events.

**Reply by jertel** [reply]
Yes, and you might also need to use `allow_buffer_time_overlap` depending on your needs. See the [related documentation](https://elastalert2.readthedocs.io/en/latest/ruletypes.html#metric-aggregation) that discusses metric aggregation with respect to time.

**Comment by nvasoft** [comment, marked as answer]
> 3. What should I use now to count?

As I expected, the calculation in the `buffer_time` window according to this expression `match["metric_count"] = match[self.rule["type"].metric_key]`
